bob/src/hw_metrics_collector.rs

```diff
@@ -179,53 +182,138 @@ enum CommandError {
-                let mut lines = output
-                    .lines()
-                    .skip_while(|line| line.find("avg-cpu:").is_none());
-                // find iowait column on the next line
-                if let Some(line) = lines.next() {
-                    let mut values = line.split_whitespace();
-                    // iowait is in 3d column
-                    if let Some(iowait_str) = values.nth(3) {
-                        if let Ok(iowait) = Self::parse_iostat_result(iowait_str) {
-                            return Ok(iowait);
+        const CPU_IOWAIT_COLUMN: usize = 5;
+        let mut err = None;
+        match Self::stat_cpu_line() {
+            Ok(line) => {
+                let parts: Vec<&str> = line.split_whitespace().collect();
+                if parts.len() > CPU_IOWAIT_COLUMN {
+                    let mut sum = 0.;
+                    let mut f_iowait = 0.;
+                    for i in 1..parts.len() {
+                        match parts[i].parse::<f64>() {
+                            Ok(val) => {
+                                sum += val;
+                                if i == CPU_IOWAIT_COLUMN {
+                                    f_iowait = val;
+                                }
+                            },
+                            Err(_) => {
+                                let msg = format!("Can't parse {}", CPU_STAT_FILE);
+                                err = Some(msg);
+                                break;
+                            }
                         }
                     }
+                    if err.is_none() {
+                        return Ok(f_iowait * 100. / sum);
+                    }
+                } else {
+                    let msg = format!("CPU stat format in {} changed", CPU_STAT_FILE);
+                    err = Some(msg);
                 }
-            }
+            },
             Err(e) => {
-                self.iostat_avl = false;
-                return Err(CommandError::Primary(e));
+                err = Some(e);
             }
         }
-        self.iostat_avl = false;
-        Err(CommandError::Primary("Failed to get iowait from iostat".to_string()))
+        self.procfs_avl = false;
+        Err(CommandError::Primary(err.unwrap()))
+    }
+}
+
+const DIFFCONTAINER_THRESHOLD: Duration = Duration::from_millis(500);
+
+struct DiffContainer<T> {
+    last: Option<T>,
+}
+
+impl<T> DiffContainer<T>
+where T: std::ops::Sub<Output = T> +
+         Copy,
+{
+    fn new() -> Self {
+        Self {
+            last: None,
+        }
+    }
+
+    fn diff(&mut self, new: T) -> Option<T> {
+        if let Some(last) = self.last {
+            let diff = new - last;
+            self.last = Some(new);
+            Some(diff)
+        } else {
+            self.last = Some(new);
+            None
+        }
+    }
+}
+
+#[derive(Clone, Copy)]
+struct DiskStats {
+    reads: u128,
+    writes: u128,
+    read_time: u128,
+    write_time: u128,
```

Comment:
Why do you use `u128`? I checked the documentation and these values are stored by the OS in `unsigned long`, which is `u64` in Rust.

https://www.kernel.org/doc/Documentation/admin-guide/iostats.rst